### ISSUE TITLE
Функция печатает массив любого типа

### ISSUE THREAD

**Issue by Geniok** [Pahtus200Maks0/Test#1]
К этой программе у меня несколько вопросов.

Начнем с этих строк:
`unsigned uArr[] = {12, 45, 667, 221, 345};`
`int iArr[] = {34, 43, 34, 456};`
`double dArr[] = {12.2, 0.45, 6.67, 22.1, 3.45};`

Если вы читали внимательно Прата, то он в своей книге пишет, что в данном случае необходимо применять квалификатор const и в целом, я с ним согласен.

`const unsigned uArr[] = {12, 45, 667, 221, 345};`
`const int iArr[] = {34, -43, -34, 456};   // Обратите внимание, что появились отрицательные числа, показывающие разницу между этим массивом и первым`
`const double dArr[] = {12.2, 0.45, 6.67, 22.1, 3.45};`

Перейдем теперь к этой строке:
`printArray(uArr, COUNT(uArr), "unsigned");`

Я бы лично данную функцию реализовал чуть иначе:
`printArray(&uArr[0], "unsigned");`

Поясню, почему так.
Во-первых, на мой взгляд не имеет смысла заставлять пользователя вашей функции самому вычислять размер массива каждый раз при вызове функции, тогда как логичнее это реализовать в самой функции.
Во-вторых, при таком вызове первого параметра, видно сразу, что в функцию передается указатель на массив. Здесь работает правило, что явное указание лучше неявного. Это не принципиально, но тот, кто будет читать вашу программу, меньше потратит времени на ее понимание.

Далее, прототип функции 
`void printArray (void *ptr, unsigned count, char *type)`
я бы переписал в вашем случае в
`void printArray (void *ptr, const unsigned count, const char *type)`
или в моем случае в
`void printArray (void *ptr,  const char *type)`

Во-первых, потому что передается указатель на строку, длинна которой не меняется.
Во-вторых, потому что если посмотреть прототип функции strcmp, то видно, что параметрами у нее являются как раз константные указатели,  а не просто указатели. То, как написано у вас, нарушает соглашение о вызове функции, обычно при правильной настройке компилятора, он выдает об этом предупреждение.

**Comment by Pahtus200Maks0** [comment]
Да, согласен. Если предполагается, что функция не должна меняет переданных ей значений, то надо использовать квалификатор const.  Ну и да, в самой функции лучше будет реализовать подсчёт размера массива, а не через макрос.

**Comment by Geniok** [comment]
> Да, согласен. Если предполагается, что функция не должна меняет переданных ей значений, то надо использовать квалификатор const. Ну и да, в самой функции лучше будет реализовать подсчёт размера массива, а не через макрос.

Я бы вам посоветовал при написании любого кода думать не со стороны программиста, которому нужно решить задачу, а со стороны пользователя, который будет пользоваться вашим ПО или вызывать ваши функции из своего кода.
Такой подход даст вам возможность сразу продумывать наиболее оптимальные варианты как интерфейсов, так и прототипы функций.

**Comment by Pahtus200Maks0** [comment]
Возьму на заметку.
